///**************************************************************************
//* @IDL Module: ZCHXAISVessel
//* @File: ZCHXAISVessel.proto
//* @Description:
//*    船舶明细和船舶标准格式;
//* @Copyright: Copyright (c) 2014
//* @Company: 深圳置辰海信科技有限公司
//* @WebSite: http://www.szcenterstar.com/
//* @author 
//* @Revision History
//*
//* <pre>
//* ----------------------------------------------------------------------
//*   Ver     Date       Who             Comments
//*  ----- ----------  --------  ---------------------------------------
//*   1.0  2015/09/30   李鹭           形成初稿
//* ----------------------------------------------------------------------
//* </pre>
//**************************************************************************/


package com.zhichenhaixin.proto;

    //----------------------------------------------------------------------------------------------------------
    // EPFD  电子定位装置类型
    //----------------------------------------------------------------------------------------------------------
    //-|num|-------------Status Discription---------|------------中文-------------------
    // | 0 | Undefined                              |  未定义(默认)
    // | 1 | GPS                                    |  GPS
    // | 2 | GLONASS                                |  GLONASS
    // | 3 | Combined GPS/GLONASS                   |  组合 GPS/GLONASS
    // | 4 | Loran-C                                |  Loran-C
    // | 5 | Chayka                                 |  Chayka
    // | 6 | Integrated navigation system           |  综合导航系统
    // | 7 | Surveyed                               |  观测
    // | 8 - 15 | Galileo                           |  不用
    //----------------------------------------------------------------------------------------------------------
    enum EPFD{
        UNDEFINED  = 0;
        GPS = 1;
        GLONASS = 2;
        COMBINED_GPS_AND_GLONASS = 3;
        LORAN_C = 4;
        CHAYKA = 5;
        INTEGRATED_NAVIGATION_SYSTEM = 6;
        SURVEYED = 7;
        GALILEO = 8;
    }

    //----------------------------------------------------------------------------------------------------------
    // 船舶静态报告
	// required uint64 timeOfDay = 21;			// 当日时间 时分秒的秒数，世界标准时间(UTC)表示的绝对时间戳   
    //----------------------------------------------------------------------------------------------------------
    message VesselInfo {
    	required  string id  = 1;				// 唯一识别码("AIS_"+shipType+"__"+用户识别码)
    	required  int32 mmsi  = 2;				// 用户识别码 
    	required  string shipType  = 3;			// 消息类型(A类，BCS类,BSO类)
    	required  int32 imo  = 4; 				// IMO 号码	
    	required  string callSign  = 5; 		// Call Sign 呼号	
    	required  string shipName  = 6; 		// 船名	
    	required  int32  cargoType  = 7; 		// 船舶类型：0其他,1渔船,2客船,3货船,4油轮,5危险品船
    	required  string country  = 8; 			// 国籍	
    	required  string vendorID  = 9;     	// Vendor ID 制造商	
    	required  int32 shipLength  = 10; 		// 船长
    	required  int32 shipWidth  = 11; 		// 船高
    	required  int32 toBow   = 12; 			// dim to a
    	required  int32 toStern  = 13;			// dim to b
    	required  int32 toPort  = 14;			// dim to c
    	required  int32 toStarboard  = 15; 		// dim to d
    	required  EPFD fixType  = 16;			// EPFD 电子定位装置类型
    	required  string eta  = 17;       		// 预计到达时间
    	required  float draught  = 18; 			// 最大吃水      numeric(6,2)
    	required  string dest  = 19;			// 目的地
    	required  uint64 UTC  = 20;				// 时间标记
   }

    //----------------------------------------------------------------------------------------------------------
    // Navigation Status  航行状态
    //----------------------------------------------------------------------------------------------------------
    //-|num|-------------Status Discription---------------------------------|------------中文-------------------
    // | 0 | Under way using engine                                         |  在航(主机推动)
    // | 1 | At anchor                                                      |  锚泊
    // | 2 | Not under command                                              |  失控
    // | 3 | Restricted manoeuverability                                    |  操纵受限
    // | 4 | Constrained by her draught                                     |  吃水受限
    // | 5 | Moored                                                         |  靠泊
    // | 6 | Aground                                                        |  搁浅
    // | 7 | Engaged in Fishing                                             |  从事捕鱼
    // | 8 | Under way sailing                                              |  靠船帆提供动力
    // | 9 | Reserved for future amendment of Navigational Status for HSC   |  为将来船舶运输 DG,HS,MP或IMO规定的有毒或污染C(HSC)类货物的航行状态修正 所保留
    // | 10 | Reserved for future amendment of Navigational Status for WIG  |  为将来船舶运输 DG,HS,MP或IMO规定的有毒或污染A(WIG)类货物的航行状态修正 所保留
    // | 11 ~ 14 | Reserved for future use                                  |  为将来使用所保留
    // | 15 | Not defined (default)                                         |  未定义(默认)
    //----------------------------------------------------------------------------------------------------------
    enum NAVI_STATUS{
        UNDER_WAY_USING_ENGINE = 0;
        AT_ANCHOR = 1;
        NOT_UNDER_COMMAND = 2;
        RESTRICTED_MANOEUVER = 3;
        CONSTRAINED_BY_DRAUGHT = 4;
        MOORED = 5;
        AGROUND = 6;
        ENGAGED_IN_FISHING = 7;
        UNDER_WAY_SAILING = 8;
        RESERVED_FOR_FUTURE_AMENDMENT_FOR_HSC = 9;
        RESERVED_FOR_FUTURE_AMENDMENT_FOR_WIG = 10;
        RESERVED_FOR_FUTURE_USE_1 = 11;
        RESERVED_FOR_FUTURE_USE_2 = 12;
        RESERVED_FOR_FUTURE_USE_3 = 13;
        RESERVED_FOR_FUTURE_USE_4 = 14;
        NOT_DEFINED = 15;
    }
	
    //----------------------------------------------------------------------------------------------------------
    // 船舶动态报告
    //----------------------------------------------------------------------------------------------------------
    message VesselTrack {
    	required string id = 1;					// 唯一识别码("AIS_"+shipType+"__"+用户识别码)
    	required int32 mmsi = 2;				// 用户识别码
    	required string shiptype = 3;			// 船舶种类(A类，BCS类,BSO类)
    	required NAVI_STATUS navStatus = 4;		// 船舶航行状态
    	required float rot = 5;					// 船舶转向率 degree/min
    	required float sog = 6;					// 对地航速 单位：节
    	required double lon = 7;				// 经度
    	required double lat = 8;				// 纬度
    	required float cog = 9;					// 对地航向
    	required float heading = 10;			// 航首向
    	required uint64 UTC = 11;				// 时间标记  
		optional  int32 type = 12;				// 类型 0：ais，1：北斗，2：CMDA 3:融合  4:融合 船讯网数据(船舶档案)
    	optional  uint32 trackNumber = 13;		// 融合雷达航迹号 I010/161

    }

    //----------------------------------------------------------------------------------------------------------
    // 基地台报告
    //----------------------------------------------------------------------------------------------------------
    message BaseStationReport {
        required int32 id = 1;                  // 唯一识别码
        required int32 repeatIndicator = 2;     // 转发指示符重发次数
        required int32 mmsi = 3;                // 用户识别码 长度 9
        required int32 year = 4;                // UTC年份
        required int32 month = 5;               // UTC月份
        required int32 day = 6;                 // UTC日期
        required int32 hour = 7;                // UTC小时
        required int32 minute = 8;              // UTC分钟
        required int32 second = 9;              // UTC秒
        required int32 positionAccuracy = 10;   // 船位精准度
        required double lon = 11;               // 经度
        required double lat = 12;               // 纬度
        required int32 fixType = 13;            // 电子定位装置类型
        required uint64 UTC = 14;               // 时间标记  
    }

   //----------------------------------------------------------------------------------------------------------
    // Aton Type  助航类型
    //----------------------------------------------------------------------------------------------------------
    //-|num|-------------Status Discription---------------------------------|------------中文-------------------
    // | 0 | Default                                                        |  未指明(默认)
    // | 1 | Reference point                                                |  参考点
    // | 2 | RACON (radar transponder marking a navigation hazard)          |  雷达信标
    // | 3 | Fixed structure off shore                                      |  离岸建筑
    // | 4 | Spare, Reserved for future use.                                |  备用(未来可用)
    // | 5 | Light, without sectors                                         |  信号灯，不带分区
    // | 6 | Light, with sectors                                            |  信号灯，带分区
    // | 7 | Leading Light Front                                            |  导航灯，前
    // | 8 | Leading Light Rear                                             |  导航灯，后
    // | 9 | Beacon, Cardinal N                                             |  信标，主北
    // | 10 | Beacon, Cardinal E                                            |  信标，主东
    // | 11 | Beacon, Cardinal S                                            |  信标，主南
    // | 12 | Beacon, Cardinal W                                            |  信标，主西
    // | 13 | Beacon, Port hand                                             |  信标，左舷
    // | 14 | Beacon, Starboard hand                                        |  信标，右舷
    // | 15 | Beacon, Preferred Channel port hand                           |  信标，推荐航道左舷
    // | 16 | Beacon, Preferred Channel starboard hand                      |  信标，推荐航道右舷
    // | 17 | Beacon, Isolated danger                                       |  信标，孤立障碍物
    // | 18 | Beacon, Safe water                                            |  信标，安全水域
    // | 19 | Beacon, Special mark                                          |  信标，特殊标志
    // | 20 | Cardinal Mark N                                               |  方向标志 北
    // | 21 | Cardinal Mark E                                               |  方向标志 东
    // | 22 | Cardinal Mark S                                               |  方向标志 南
    // | 23 | Cardinal Mark W                                               |  方向标志 西
    // | 24 | Port hand Mark                                                |  左舷标志
    // | 25 | Starboard hand Mark                                           |  右舷标志
    // | 26 | Preferred Channel Port hand                                   |  推荐航道左舷标志
    // | 27 | Preferred Channel Starboard hand                              |  推荐航道右舷标志
    // | 28 | Isolated danger                                               |  孤立障碍物
    // | 29 | Safe Water                                                    |  安全水域
    // | 30 | Special Mark                                                  |  特殊标志
    // | 31 | Light Vessel / LANBY / Rigs                                   |  灯船/LANBY/无线电
    //----------------------------------------------------------------------------------------------------------
   enum ATON_TYPE{
        DEFAULT = 0;
        REFERENCE_POINT  = 1;
        RACON = 2;
        FIXED_STRUCTURE_OFF_SHORE = 3;
        SPARE = 4;
        LIGHT_WITHOUT_SECTORS  = 5;
        LIGHT_WITH_SECTORS  = 6;
        LEADING_LIGHT_FRONT = 7;
        LEADING_LIGHT_REAR = 8;
        BEACON_CARDINAL_N  = 9;
        BEACON_CARDINAL_E = 10;
        BEACON_CARDINAL_S  = 11;
        BEACON_CARDINAL_W = 12;
        BEACON_PORT_HAND = 13;
        BEACON_STARBOARD_HAND = 14;
        BEACON_PREFERRED_CHANNEL_PORT_HAND  = 15;
        BEACON_PREFERRED_CHANNEL_STARBOARD_HAND  = 16;
        BEACON_ISOLATED_DANGER  = 17;
        BEACON_SAFE_WATER  = 18;
        BEACON_SPECIAL_MARK  = 19;
        CARDINAL_MARK_N   = 20;
        CARDINAL_MARK_E  = 21;
        CARDINAL_MARK_S  = 22;
        CARDINAL_MARK_W  = 23;
        PORT_HAND_MARK  = 24;
        STARBOARD_HAND_MARK  = 25;
        PREFERRED_CHANNEL_PORT_HAND  = 26;
        PREFERRED_CHANNEL_STARBOARD_HAND   = 27;
        ISOLATED_DANGER  = 28;
        SAFE_WATER   = 29;
        SPECIAL_MARK  = 30;
        LIGHT_VESSEL_LANBY_RIGS  = 31;
    }

    //----------------------------------------------------------------------------------------------------------
    // 助航设备报告
    //----------------------------------------------------------------------------------------------------------
    message AidtoNavigationReport {
        required int32 id = 1;                  // 唯一识别码
        required int32 repeatIndicator = 2;     // 转发指示符重发次数
        required int32 mmsi = 3;                // 用户识别码 长度 9
        required ATON_TYPE atonType = 4;        // 助航类型
        required string name = 5;               // 助航名称
        required int32 positionAccuracy = 6;    //船位精准度
        required double lon = 7;                //经度
        required double lat = 8;                //纬度
        required int32 toBow   = 9;           // dim to a
        required int32 toStern  = 10;          // dim to b
        required int32 toPort  = 11;           // dim to c
        required int32 toStarboard  = 12;      // dim to d
        required int32 fixType = 13;            //电子定位装置类型
        required uint64 UTC = 14;               // 时间标记  
    }


    message AIS {
        required int32 flag = 1 ;   // 消息类型
        required string sourceId = 2; //数据来源唯一识别符
        optional VesselInfo   vesselInfo = 3;
        optional VesselTrack  vesselTrack = 4;
        optional BaseStationReport  baseStationReport = 5;
        optional AidtoNavigationReport  aidtoNavigationReport = 6;
        
    }
    
	//----------------------------------------------------------------------------------------------------------
    // 接收采集器ais实时数据信息
    //----------------------------------------------------------------------------------------------------------
	message AISList{
		required uint64 UTC = 1;
		repeated AIS ais = 2;
	}
	
	//基站数据示意图
	message AisChart
	{
		required uint64 id = 1;                    // 基站ID
        required string name = 2;                  // 基站名称
        required double latitude = 3;			        // 基站 WGS84坐标计算纬度 I010/041
    	required double longitude = 4;			        // 基站 WGS84坐标计算经度 I010/041
        required uint64 UTC = 5;                        // 时间标记  
        required bytes imageData  = 6;                 // 图像数据
        required uint32 width = 7;                      // 图像宽度
        required uint32 height = 8;                     // 图像高度
        required double radius = 9;			            // 图片对应的半径大小
		required string format = 10;					// 图片的格式 png 或者svg
	}
